### PR TITLE
mgr/dashboard: fix NVMeoF API

### PR DESCRIPTION
**Pull request by epuertat**
[REST API Design doc.](https://docs.google.com/document/d/13zjDcdAzrWBMhODVkvAOb1htJ5uJBsWLbo3oXOaYRio/edit)

* Update NVMe-oF gRPC Proto to 1.0.0
* Error handling,
* Missing PATCH for certain namespace ops (resize, set QoS, set balance
  groups),
* Stop bypassing gRPC payloads and validate those in the back-end,
* Fix incorrect HTTP 1.1 semantics for some POST/DELETE and URIs.
* Catch errors/exceptions.
* Clean-up EndpointDoc Params
* Run Black linter.
* Remove most of NVMeoFClient glue code between gRPC and controller.
* Fix namespace delete endpoint by exposing trsvcid
* nvmeof io_stats support

Fixes: https://tracker.ceph.com/issues/64890

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by nizamial09 — COMMENTED** [review]
tested this.

looks like all the endpoints are working perfectly. :tada: 

Although the namespace update failed for me with
```
Mar 20 19:49:41 ceph-node-00 ceph-mgr[2841]: [dashboard INFO nvmeof_client] Found nvmeof gateway at 192.168.100.101:5500
Mar 20 19:49:41 ceph-node-00 ceph-mgr[2841]: [dashboard ERROR exception] Dashboard Exception
                                             Traceback (most recent call last):
                                               File "/usr/share/ceph/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
                                                 return handler(*args, **kwargs)
                                               File "/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 54, in __call__
                                                 return self.callable(*self.args, **self.kwargs)
                                               File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 263, in inner
                                                 ret = func(*args, **kwargs)
                                               File "/usr/share/ceph/mgr/dashboard/controllers/_rest_controller.py", line 193, in wrapper
                                                 return func(*vpath, **params)
                                               File "/usr/share/ceph/mgr/dashboard/services/nvmeof_client.py", line 275, in wrapper
                                                 func(*args, **kwargs)
                                               File "/usr/share/ceph/mgr/dashboard/services/nvmeof_client.py", line 268, in wrapper
                                                 component='nvmeof')
                                             dashboard.exceptions.DashboardException: Failure resizing namespace: Failure resizing bdev bdev_f87fa4c0-1220-484d-9dc1-2baa1d2cf0ae: Invalid argument
```

I am sure it might be something I am doing wrong. The below was the req
```
curl -X 'PATCH' \
  'https://192.168.31.207:4200/api/nvmeof/subsystem/nqn.2016-06.io.spdk%3Acnode1/namespace/1' \
  -H 'accept: application/vnd.ceph.api.v1.0+json' \
  -H 'Content-Type: application/json' \
  -d '{
  "uuid": null,
  "rbd_image_size": 3,
  "load_balancing_group": null,
  "rw_ios_per_second": null,
  "rw_mbytes_per_second": null,
  "r_mbytes_per_second": null,
  "w_mbytes_per_second": null
}'
```

BTW, I ended up testing this in cephadm. I had issues with the make scripts which I wasn't able to fix so I switched to cephadm.

**Comment by nizamial09** [comment]
while testing this, I realized we may have to mark these endpoints as EXPERIMENTAL? wdyt?

**Comment by nizamial09** [comment]
btw, just saw that the `GET /api/hosts` endpoint is failing with a weird error
```
curl -X 'GET' \
  'https://192.168.100.100:8443/api/host?facts=false&offset=0&limit=5' \
  -H 'accept: application/vnd.ceph.api.v1.3+json'
```
and the response is

```
{
  "status": "415 Unsupported Media Type",
  "detail": "Incorrect version: endpoint is '1.0', client requested '1.3'",
  "request_id": "cbd3fa63-1596-4d71-a5ce-dd0d58c7b1a2"
}
```

The version of that endpoint by default is 1.3

**Comment by epuertat** [comment]
> Although the namespace update failed for me with
> 
> ```
> Mar 20 19:49:41 ceph-node-00 ceph-mgr[2841]: [dashboard INFO nvmeof_client] Found nvmeof gateway at 192.168.100.101:5500
> Mar 20 19:49:41 ceph-node-00 ceph-mgr[2841]: [dashboard ERROR exception] Dashboard Exception
>                                              Traceback (most recent call last):
>                                                File "/usr/share/ceph/mgr/dashboard/services/exception.py", line 47, in dashboard_exception_handler
>                                                  return handler(*args, **kwargs)
>                                                File "/lib/python3.6/site-packages/cherrypy/_cpdispatch.py", line 54, in __call__
>                                                  return self.callable(*self.args, **self.kwargs)
>                                                File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 263, in inner
>                                                  ret = func(*args, **kwargs)
>                                                File "/usr/share/ceph/mgr/dashboard/controllers/_rest_controller.py", line 193, in wrapper
>                                                  return func(*vpath, **params)
>                                                File "/usr/share/ceph/mgr/dashboard/services/nvmeof_client.py", line 275, in wrapper
>                                                  func(*args, **kwargs)
>                                                File "/usr/share/ceph/mgr/dashboard/services/nvmeof_client.py", line 268, in wrapper
>                                                  component='nvmeof')
>                                              dashboard.exceptions.DashboardException: Failure resizing namespace: Failure resizing bdev bdev_f87fa4c0-1220-484d-9dc1-2baa1d2cf0ae: Invalid argument
> ```
> 
> I am sure it might be something I am doing wrong. The below was the req
> 
> ```
> curl -X 'PATCH' \
>   'https://192.168.31.207:4200/api/nvmeof/subsystem/nqn.2016-06.io.spdk%3Acnode1/namespace/1' \
>   -H 'accept: application/vnd.ceph.api.v1.0+json' \
>   -H 'Content-Type: application/json' \
>   -d '{
>   "uuid": null,
>   "rbd_image_size": 3,
>   "load_balancing_group": null,
>   "rw_ios_per_second": null,
>   "rw_mbytes_per_second": null,
>   "r_mbytes_per_second": null,
>   "w_mbytes_per_second": null
> }'
> ```

@nizamial09 what was the initial size of the rbd image? Size 3 here means 3 bytes, but internally Dashboard will round it up to the next multiple of MiB (1024*1024 bytes), by the expression `int((rbd_image_size + mib - 1) / mib)`, which in this very case gives `1` (MiB).

However, if the original image was 1 MiB or larger, then resize will fail (the error you see there directly comes from SPDK JSON RPC).

> BTW, I ended up testing this in cephadm. I had issues with the make scripts which I wasn't able to fix so I switched to cephadm.

@adk3798 do you know if the `nvmeof` service is deployed with RBD write permissions?

**Comment by epuertat** [comment]
> while testing this, I realized we may have to version these endpoints as EXPERIMENTAL? wdyt?

@nizamial09 I also thought that, but as we don't want to change that (it'd break the plugin as soon as those endpoints are marked as non-experimental), I went for default/1.0 versioning.

**Comment by nizamial09** [comment]
> what was the initial size of the rbd image? Size 3 here means 3 bytes, but internally Dashboard will round it up to the next multiple of MiB (1024*1024 bytes), by the expression `int((rbd_image_size + mib - 1) / mib)`, which in this very case gives `1` (MiB).

Ah okay, so bytes was the unit. i was a bit confused here. Can we mention the unit in the Endpoint doc? 

> what was the initial size of the rbd image?

@epuertat It was 1 GiB

**Comment by epuertat** [comment]
> > what was the initial size of the rbd image? Size 3 here means 3 bytes, but internally Dashboard will round it up to the next multiple of MiB (1024*1024 bytes), by the expression `int((rbd_image_size + mib - 1) / mib)`, which in this very case gives `1` (MiB).
> 
> Ah okay, so bytes was the unit. i was a bit confused here. Can we mention the unit in the Endpoint doc?
> 
> > what was the initial size of the rbd image?
> 
> @epuertat It was 1 GiB

Ok @nizamial09 in that case it's expected failure. You can shrink RBD images with NVMEoF.

File: src/pybind/mgr/dashboard/controllers/nvmeof.py [hunk 1]
```diff
@@ -1,344 +1,337 @@
...
-        "List all NVMeoF namespaces",
-        parameters={"nqn": Param(str, "NVMeoF subsystem NQN")},
-    )
-    @map_collection(model.Namespace, pick="namespaces")
-    @handle_nvmeof_error
-    def list(self, nqn: str):
-        return NVMeoFClient().stub.list_namespaces(
-            NVMeoFClient.pb2.list_namespaces_req(subsystem=nqn)
+try:
+    from ..services.nvmeof_client import NVMeoFClient, empty_response, \
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/controllers/nvmeof.py:10]: huh, tox fix command modified this one

**Comment by nizamial09** [comment]
@epuertat i pushed a commit here after testing. All the endpoints were working except the `namespace delete` which I fixed here by adding `trsvcid` to the request. Fixed some test failures too.

I also bought the try-except block because if the grpc is not available, the whole `NVMeoFClient` is not gonna be there and that broke the import in the controller.

I didn't bumped nvmeof to 1.1.0 because locally I bumped into the below error when starting the gateway
```
2024-03-21T15:33:50.712+0000 7f9a0d1875c0  0 nvmeofgw int NVMeofGwMonitorClient::init() nvmeof Registered monc callback
2024-03-21T15:33:50.714+0000 7f9a0d1875c0  0 nvmeofgw int NVMeofGwMonitorClient::init() monc.authentication done
2024-03-21T15:33:50.715+0000 7f9a0d1875c0  0 nvmeofgw void NVMeofGwMonitorClient::tick() 
2024-03-21T15:33:50.716+0000 7f9a0d1875c0  0 nvmeofgw bool get_gw_state(const char*, const std::map<std::pair<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char> >, std::map<std::__cxx11::basic_string<char>, NvmeGwState> >&, const NvmeGroupKey&, const NvmeGwId&, NvmeGwState&) can not find group (rbd,) old map map: {}
2024-03-21T15:33:50.716+0000 7f9a0d1875c0  0 nvmeofgw void NVMeofGwMonitorClient::send_beacon() sending beacon as gid 24538 availability 0 osdmap_epoch 0
2024-03-21T15:33:50.716+0000 7f9a0d1875c0  0 nvmeofgw int NVMeofGwMonitorClient::init() Complete.
2024-03-21T15:33:50.716+0000 7f9a09124640  0 nvmeofgw virtual bool NVMeofGwMonitorClient::ms_dispatch2(ceph::ref_t<Message>&) got map type 4
2024-03-21T15:33:50.716+0000 7f9a09124640  0 ms_deliver_dispatch: unhandled message 0x56386a3701a0 mon_map magic: 0 v1 from mon.1 v2:192.168.100.101:3300/0
2024-03-21T15:33:50.716+0000 7f9a09124640  0 nvmeofgw virtual bool NVMeofGwMonitorClient::ms_dispatch2(ceph::ref_t<Message>&) got map type 41
2024-03-21T15:33:50.716+0000 7f9a09124640  0 client.0 ms_handle_reset on v2:192.168.100.101:3300/0
2024-03-21T15:33:50.716+0000 7f9a09124640  0 client.0 ms_handle_reset on v2:192.168.100.101:3300/0
2024-03-21T15:33:50.717+0000 7f9a09124640  0 nvmeofgw virtual bool NVMeofGwMonitorClient::ms_dispatch2(ceph::ref_t<Message>&) got map type 4
2024-03-21T15:33:50.717+0000 7f9a09124640  0 ms_deliver_dispatch: unhandled message 0x56386a3704e0 mon_map magic: 0 v1 from mon.1 v2:192.168.100.101:3300/0
```

this error was in a loop and logged continously

```
NVMeofGwMonitorClient::ms_dispatch2(ceph::ref_t<Message>&) got map type 4
2024-03-21T15:33:50.717+0000 7f9a09124640  0 ms_deliver_dispatch: unhandled message 0x56386a3704e0 mon_map magic: 0 v1 from mon.1 v2:192.168.100.101:3300/0
```

**Comment by epuertat** [comment]
jenkins retest this please

**Comment by nizamial09** [comment]
squashed the commits.

**Review by nizamial09 — APPROVED** [review]

**Comment by nizamial09** [comment]
fixed or suppressed [lint error](https://jenkins.ceph.com/job/ceph-pull-requests/131943/testReport/junit/projectroot.src.pybind.mgr/dashboard/run_tox_mgr_dashboard_lint/)
```
[NON-XML-CHAR-0x1B][0;36m./model/nvmeof.py[NON-XML-CHAR-0x1B][0m:[NON-XML-CHAR-0x1B][1;33m1[NON-XML-CHAR-0x1B][0m:[NON-XML-CHAR-0x1B][1;32m1[NON-XML-CHAR-0x1B][0m: [NON-XML-CHAR-0x1B][1;31mF401[NON-XML-CHAR-0x1B][0m 'typing.List' imported but unused
1     E501 line too long (110 > 100 characters)
1     F401 'typing.List' imported but unused
lint: exit 1 (8.85 seconds) /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard> flake8 pid=2271084
  lint: FAIL code 1 (76.70=setup[67.86]+cmd[8.85] seconds)
  evaluation failed :( (76.83 seconds)
```

**Review by epuertat — COMMENTED** [review]
I can't approve my own PR but your changes @nizamial09 LGTM! 👍🏼

**Comment by epuertat** [comment]
jenkins test dashboard cephadm

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
unrelated [e2e failure](https://jenkins.ceph.com/job/ceph-dashboard-pull-requests/14657/consoleFull#-208797045573a8703b-5adb-41c5-84a0-8cf4e065ba3d).
```
  Running:  ui/notification.e2e-spec.ts                                                   (29 of 33)
  Estimated: 1 minute, 10 seconds


  Notification page
    1) "before all" hook for "should open notification sidebar"
    2) "after all" hook for "should open notification sidebar"


  0 passing (2m)
  2 failing

  1) Notification page
       "before all" hook for "should open notification sidebar":
     AssertionError: Timed out retrying after 120000ms: Expected to find element: `input[name=pgNum]`, but never found it.

Because this error occurred during a `before all` hook we are skipping the remaining tests in the current suite: `Notification page`
```

**Pull request merged**